### ISSUE TITLE
Allow easybuild to set global configuration variables

### ISSUE THREAD

**Issue by SimonPinches** [hpcugent/easybuild-framework#1689]
Allow git-like global configuration for things like github-user.

Something equivalent to git config --global set.user xxx.yyy

**Comment by wpoely86** [comment]
What do you mean? We have a global configuration file?

**Comment by boegel** [comment]
The idea is to support changing the configuration from the `eb` command line:

Something like:
```
eb --edit-config github-user boegel
```

I kind of like the idea, since EasyBuild could verify what you're providing before actually applying it (e.g. check that the GitHub user exists).

**Comment by wpoely86** [comment]
I don't agree. We can have a verify configuration file thingie but let EB write to it seems like a bad idea to me. The configuration file should not be touched by any program.

**Comment by SimonPinches** [comment]
It was just a suggestion since it works well for [`git`](https://git-scm.com/docs/git-config) which allows you to query/set/replace/unset configuration options all from the command line.

**Comment by boegel** [comment]
@SimonPinches To be honest, I never used the `git config` subcommand (at least not that I can remember)...

Are there any big advantages of using `git config` over just editing the configuration manually?

Some I can think of include:

* avoiding introducing syntax errors
* opportunity to validate the values being provided
* not having to figure out where the config file is

None of these are major by themselves, but together they may be enough to consider supporting `eb --edit-config`...

**Comment by RalphTheBest** [comment]
Just a humble comment. I'm all in with the idea of modifying config files via command line since it will be easier to tweak certain parameters and has all the advantages said in the previous comment. 
However this arise a question (for me at least): how do you target the config file to be modified? I mean, in the case that you have two or more config files, (for different levels of configuration) which one is gonna be modified? 

The first two ideas that pop in my mind are:

1. Always writing the one with higher preference. This way is easier, but assumes that only user-level configuration can be modified. (assuming for example that we have a config for system level and another one for user level)

2. Allow the user to especify the config file to be writen. This one gives more control and flexibility, but makes the implementation harder and probably will force the user to know where the config file is.

**Comment by boegel** [comment]
Good point... How does `git` handle that, or does it only have a single config file?

**Comment by RalphTheBest** [comment]
AFAIK, git uses a param like: `git config [<file-option>]` where `file-option` can be `-global`, `-system` or `-local` (which is the default). Also, it can be `-f <filename>` if you want to specify the file to be writen.

The first three options target the default config files depending on the level of the repo you want to configure, but since Easybuild doesn't have files like that by default (or that's what I think) the same options cannot be applied.

A possible solution could be targeting always the config file with higher preference by default and include something like `-f <filename>` when you know which file you want to modify. This way, if the user doesn't know (or doesn't want to know) where the config file is, it's possible to just target the one that it's gonna ensure the changes. If for any reason you want to modify other file than the previous one, it probably means that you know the location. 

Sounds reasonable?

**Comment by boegel** [comment]
@RalphTheBest sounds like a good design to me, indeed... Now, who's going to implement it? ;)

**Comment by RalphTheBest** [comment]
I'm not very familiar with the EB code or the way it works, so it will take me some (probably a lot of) time. But since this doesn't seem like an urgent feature (and if no one minds waiting until I decipher all the mysteries EB has for me) I think I can try to tackle it.

Anyway, if you could point me the places where I have to look first and where should the changes go, it would be really useful. 

For what I've seen, adding some lines in the [easybuild/tools/options.py](https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/tools/options.py) is necessary. Where should I implement the option itself? Any other thing I have to pay attention to?

**Comment by boegel** [comment]
@RalphTheBest The actual implementation should probably go into a new function in `easybuild/tools/config.py` .

If you add a new option, having a test for it would be nice, but I can help there.

**Comment by RalphTheBest** [comment]
@boegel I'll start working on it as soon as I have some time. When I have something worthy of a PR we can talk about the test part, I have no problem with adding it, as long as you don't mind some questions in the process. (The less I screw working on it, the less to be fixed later) ;)

**Comment by boegel** [comment]
@RalphTheBest don't hesitate to ask questions! Our IRC channel is particularly useful for that (`#easybuild` on the Freenode IRC network)
